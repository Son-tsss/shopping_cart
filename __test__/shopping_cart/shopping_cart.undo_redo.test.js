import ShoppingCart from '../../src/shopping_cart'

it('should undo the last changes', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.undo()

    const real = shoppingCart.historyState

    expect(real).toMatchSnapshot()
})

it('undo shouldn`t change anything if there is no previous states', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.undo()
    shoppingCart.undo()
    shoppingCart.undo()
    shoppingCart.undo()

    const real = shoppingCart.historyState

    expect(real).toMatchSnapshot()
})

it('canUndo should return true if there are previous states', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)

    const real = shoppingCart.canUndo()
    const expected = true

    expect(real).toEqual(expected)
})

it('canUndo should return false if there is no previous states', ()=>{
    const shoppingCart = new ShoppingCart

    shoppingCart.undo()

    const real = shoppingCart.canUndo()
    const expected = false
    expect(real).toEqual(expected)
})

it('should redo the last changes', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.undo()
    shoppingCart.redo()

    const real = shoppingCart.historyState

    expect(real).toMatchSnapshot()
})

it('redo shouldn`t change anything if there is no future states', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.redo()

    const real = shoppingCart.historyState

    expect(real).toMatchSnapshot()
})

it('canRedo should return true if there are future states', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.undo()

    const real = shoppingCart.canRedo()
    const expected = true

    expect(real).toEqual(expected)
})

it('canUndo should return false if there is no future states', ()=>{
    const shoppingCart = new ShoppingCart

    const real = shoppingCart.canRedo()
    const expected = false
    expect(real).toEqual(expected)
})

it('should empty future states if there are new changes made', ()=>{
    const shoppingCart = new ShoppingCart

    const newItem  = {
        id: "i1",
        price: 100
    }

    shoppingCart.addItem(newItem)
    shoppingCart.incrementItem(newItem.id)
    shoppingCart.undo()
    shoppingCart.incrementItem(newItem.id)

    const real = shoppingCart.canRedo()
    const expected = false

    expect(real).toEqual(expected)
})