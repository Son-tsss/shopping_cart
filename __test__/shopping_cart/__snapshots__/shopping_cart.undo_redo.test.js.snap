// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo shouldn\`t change anything if there is no future states 1`] = `
Object {
  "currentState": Object {
    "discount": 1,
    "items": Array [
      Object {
        "count": 2,
        "discount": 1,
        "id": "i1",
        "price": 100,
      },
    ],
  },
  "futureStates": Array [],
  "previousStates": Array [
    Object {
      "discount": 1,
      "items": Array [],
    },
    Object {
      "discount": 1,
      "items": Array [],
    },
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 1,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
  ],
}
`;

exports[`should redo the last changes 1`] = `
Object {
  "currentState": Object {
    "discount": 1,
    "items": Array [
      Object {
        "count": 3,
        "discount": 1,
        "id": "i1",
        "price": 100,
      },
    ],
  },
  "futureStates": Array [],
  "previousStates": Array [
    Object {
      "discount": 1,
      "items": Array [],
    },
    Object {
      "discount": 1,
      "items": Array [],
    },
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 1,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 2,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
  ],
}
`;

exports[`should undo the last changes 1`] = `
Object {
  "currentState": Object {
    "discount": 1,
    "items": Array [
      Object {
        "count": 2,
        "discount": 1,
        "id": "i1",
        "price": 100,
      },
    ],
  },
  "futureStates": Array [
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 3,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
  ],
  "previousStates": Array [
    Object {
      "discount": 1,
      "items": Array [],
    },
    Object {
      "discount": 1,
      "items": Array [],
    },
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 1,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
  ],
}
`;

exports[`undo shouldn\`t change anything if there is no previous states 1`] = `
Object {
  "currentState": Object {
    "discount": 1,
    "items": Array [],
  },
  "futureStates": Array [
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 2,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
    Object {
      "discount": 1,
      "items": Array [
        Object {
          "count": 1,
          "discount": 1,
          "id": "i1",
          "price": 100,
        },
      ],
    },
    Object {
      "discount": 1,
      "items": Array [],
    },
  ],
  "previousStates": Array [],
}
`;
